reconcile/utils/aws_api_typed/api.py

```diff
@@ -0,0 +1,194 @@
+            aws_secret_access_key = {self.secret_access_key}
+            region = {self.region}
+            """
+        )
+
+    def build_session(self) -> Session:
+        return Session(
+            aws_access_key_id=self.access_key_id,
+            aws_secret_access_key=self.secret_access_key,
+            region_name=self.region,
+        )
+
+
+class AWSTemporaryCredentials(AWSStaticCredentials):
+    """
+    A model representing temporary AWS credentials.
+    """
+
+    session_token: str
+
+    def as_env_vars(self) -> dict[str, str]:
+        env_vars = super().as_env_vars()
+        env_vars["AWS_SESSION_TOKEN"] = self.session_token
+        return env_vars
+
+    def as_credentials_file(self, profile_name: str = "default") -> str:
+        return textwrap.dedent(
+            f"""\
+            [{profile_name}]
+            aws_access_key_id = {self.access_key_id}
+            aws_secret_access_key = {self.secret_access_key}
+            aws_session_token = {self.session_token}
+            region = {self.region}
+            """
+        )
+
+    def build_session(self) -> Session:
+        return Session(
+            aws_access_key_id=self.access_key_id,
+            aws_secret_access_key=self.secret_access_key,
+            aws_session_token=self.session_token,
+            region_name=self.region,
+        )
+
+
+class AWSApi:
+    def __init__(self, aws_credentials: AWSCredentials) -> None:
+        self.session = aws_credentials.build_session()
+        self._session_clients: list[BaseClient] = []
+        self._api_cache: dict = {}
+
+    def __enter__(self) -> AWSApi:
+        return self
+
+    def __exit__(self, *exec: Any) -> None:
+        self.close()
+
+    def close(self) -> None:
+        """Close all clients created by this API instance."""
+        for client in self._session_clients:
+            client.close()
+        self._session_clients = []
+
+    def _init_or_get_sub_api(self, api_cls: type[SubApi]) -> SubApi:
+        """Return a new or cached sub api client."""
+        if str(api_cls) in self._api_cache:
+            return self._api_cache[str(api_cls)]
+
+        match api_cls:
+            case reconcile.utils.aws_api_typed.iam.AWSApiIam:
+                client = self.session.client("iam")
+                api = api_cls(client)  # type: ignore # mypy bug, it doesn't recognize that api_cls is callable
+            case reconcile.utils.aws_api_typed.organization.AWSApiOrganizations:
+                client = self.session.client("organizations")
+                api = api_cls(client)
+            case reconcile.utils.aws_api_typed.sts.AWSApiSts:
+                client = self.session.client("sts")
+                api = api_cls(client)
+            case _:
+                raise ValueError(f"Unknown API class: {api_cls}")
+
+        self._api_cache[str(api_cls)] = api
+        self._session_clients.append(client)
+        return api
+
+    @property
```

Comment:
how about use [cached_property](https://docs.python.org/3/library/functools.html#functools.cached_property), then we don't need to maintain `_api_cache` ourselves?

Reply:
for sure, applied in [f97e742](https://github.com/app-sre/qontract-reconcile/pull/4123/commits/f97e742877fa0aef8269dd4cffa129f0bbbdbe39)

Reply:
`_init_or_get_sub_api` may need a rename since it's always `init` now

Reply:
Sure 🤦